File: 000-Mathematics/008.primeFactors.py
import math

def checkPrimes(n):
    if n == 1:
        return False
    if n == 2 or n == 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i+2) == 0:
            return False
        i += 6
    
    return True


def primeFactorsN(n):
    for i in range(2,math.ceil(math.sqrt(n))):
        if checkPrimes(i) and n % i == 0:
            while n % i == 0:
                print(i,end=' ')
                n = n // i
    if n != 1:
        print(n,end=' ')
    print()
primeFactorsN(84)


# Optimised

def primeFactors(n):
    if n <= 1:
        return
    while n % 2 == 0:
        print(2,end=' ')
        n = n // 2
    while n % 3 == 0:
        print(3,end=' ')
        n = n // 3
    i = 5
    while i * i <= n:
        while n % i == 0:
            print(i,end=' ')
            n = n// i
        while n % (i+2) == 0:
            print(i+2,end=' ')
            n = n// (i+2)
    print()
    if n > 3:
        print(n)

primeFactors(1000)

File: GeeksForGeeks/DS-Course/000-Mathematics/008.primeFactors.js
function checkPrimes(n) {
  if (n == 1) {
    return false;
  }
  if (n == 2 || n == 3) {
    return true;
  }
  if (n % 2 == 0 || n % 3 == 0) {
    return false;
  }
  let i = 5;
  while (i * i < n) {
    if (n % i == 0 || n % (i + 2) == 0) {
      return false;
    }
    i += 6;
  }
  return true;
}

function primeFactorsN(n) {
  const primeFactors = [];
  for (let i = 2; i <= parseInt(Math.sqrt(n)); i++) {
    if (checkPrimes(i) && n % i == 0) {
      while (n % i == 0) {
        primeFactors.push(i);
        n /= i;
      }
    }
  }
  if (n != 1) {
    primeFactors.push(n);
  }
  console.log(primeFactors);
}

primeFactorsN(96);
